compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -73,7 +73,7 @@ pub trait TypeFoldable<'tcx>: fmt::Debug + Clone {
     }
 
     fn has_type_flags(&self, flags: TypeFlags) -> bool {
-        self.visit_with(&mut HasTypeFlagsVisitor { flags }).is_break()
+        self.visit_with(&mut HasTypeFlagsVisitor { flags }).break_value() == Some(FoundFlags)
```

Comment:
I don't see the advantage of using a type other than `()` for the break in this case.

Reply:
(This was https://github.com/rust-lang/rust/pull/78182#discussion_r509272130)

Reply:
Oh right, I remember that comment. @lcnr can you elaborate on the motivation? We should encode it in documentation/comments

Reply:
Using a ZST we explicitly document the reasons we stop visiting and the reason we check if we terminated early.

While it isn't too relevant, it makes it easier to prevent bugs like stopping early because of a different reason and forgetting to update a call site. I also feel like using a ZST here has a very low cost and removes some questions on why we are using `is_break` and under which conditions we want to stop visiting.

Don't really care too much about this though

Reply:
I have nothing to argue against that :laughing: 

If we can manage to remove `is_break` in that case (when all visitors have been migrated), then that would enforce this by there not being a way around it.

Reply:
Oh... now that I finished reading the PR, I think we should also stop having `()` as a default for `BreakTy` and require the user to specify one (or use `!` as the default?)

Reply:
using `!` as a default would be interesting :thinking: seems just as good as not having one at all

Reply:
65cdc21 🙂